### PR TITLE
Pass down column stats aggregation info into TableWriterNode

### PR DESCRIPTION
**Pull request by kewang1024**
```
== NO RELEASE NOTE ==
```

TODO:

- [ ] Add e2e test once [velox side change](https://github.com/facebookincubator/velox/pull/5554) is merged

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 1]
```diff
@@ -1996,18 +1996,76 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
-      {node->rowCountVariable,
-       node->fragmentVariable,
-       node->tableCommitContextVariable});
+  std::vector<protocol::VariableReferenceExpression> variables = {
+      node->rowCountVariable,
+      node->fragmentVariable,
+      node->tableCommitContextVariable};
+
+  std::shared_ptr<core::AggregationNode> aggregationNode;
+  if (node->statisticsAggregation) {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2005]:
```
if (node->statisticsAggregation != nullptr) {
```
In which case statisticsAggregation is not set from coordinator?

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 2]
```diff
@@ -1996,18 +1996,76 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
+    auto outputVariables = node->statisticsAggregation->outputVariables;
+    auto aggregationMap = node->statisticsAggregation->aggregations;
+    auto groupingVariables = node->statisticsAggregation->groupingVariables;
+    VELOX_CHECK_EQ(
+        aggregationMap.size(),
+        outputVariables.size(),
+        "TableWriterNode's aggregationMap and outputVariables should be the same size");
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2012]:
```
s/aggregationMap/aggregations/
```

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 3]
```diff
@@ -1996,18 +1996,76 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
+    VELOX_CHECK_GT(
+        outputVariables.size(),
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2014]:
```
nit: !outputVariables.empty()
```

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 4]
```diff
@@ -1996,18 +1996,76 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
+    std::vector<core::AggregationNode::Aggregate> aggregates;
+    std::vector<core::FieldAccessTypedExprPtr> aggrMasks;
+    aggregateNames.reserve(aggregationMap.size());
+    aggregates.reserve(aggregationMap.size());
+    for (const auto& entry : outputVariables) {
+      aggregateNames.emplace_back(entry.name);
+      core::AggregationNode::Aggregate aggregate;
+      aggregate.call = std::dynamic_pointer_cast<const core::CallTypedExpr>(
+          exprConverter_.toVeloxExpr(aggregationMap.at(entry).call));
+      if (aggregationMap.at(entry).mask != nullptr) {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2033]: Not sure if we can we simply check mask must not be set?

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 5]
```diff
@@ -1996,18 +1996,76 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
+        aggregate.mask =
+            exprConverter_.toVeloxExpr(aggregationMap.at(entry).mask);
+      }
+      if (aggregationMap.at(entry).orderBy) {
+        for (const auto& orderBy : aggregationMap.at(entry).orderBy->orderBy) {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2038]: Not sure if we can just check order by is not set for column stats aggregation? Is it used in Java side?

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 6]
```diff
@@ -1996,18 +1996,76 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
+              exprConverter_.toVeloxExpr(orderBy.variable));
+          aggregate.sortingOrders.emplace_back(
+              toVeloxSortOrder(orderBy.sortOrder));
+        }
+      }
+      aggregates.emplace_back(aggregate);
+    }
 
+    aggregationNode = std::make_shared<core::AggregationNode>(
+        // Use the ID of the DistinctLimit plan node here to propagate the
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2049]: DistinctLimit?

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 7]
```diff
@@ -1996,18 +1996,76 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
+        // stats.
+        node->id,
+        core::AggregationNode::Step::kPartial,
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2052]: Is it final here given the new design? Thanks!

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 8]
```diff
@@ -1996,18 +1996,76 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
-  auto outputType = toRowType(
-      {node->rowCountVariable,
-       node->fragmentVariable,
-       node->tableCommitContextVariable});
+  std::vector<protocol::VariableReferenceExpression> variables = {
+      node->rowCountVariable,
+      node->fragmentVariable,
+      node->tableCommitContextVariable};
+
+  std::shared_ptr<core::AggregationNode> aggregationNode;
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2004]: Can you provide a utility say node = buildStatsAggregationNode(protocol node, &variables)?

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 9]
```diff
@@ -1996,18 +1996,76 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
   auto insertTableHandle =
       std::make_shared<core::InsertTableHandle>(connectorId, hiveTableHandle);
 
-  auto outputType = toRowType(
-      {node->rowCountVariable,
-       node->fragmentVariable,
-       node->tableCommitContextVariable});
+  std::vector<protocol::VariableReferenceExpression> variables = {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:1999]:
```
s/variables/outputVariables/
```

**Review by xiaoxmeng — COMMENTED** [review]
@kewang1024 thanks for the change!

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp (hunk 4 above)
Comment (kewang1024) [reply presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2033]: This is the same logic we get from converting AggregationNode, I just merged those logic into one utility function

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp (hunk 7 above)
Comment (kewang1024) [reply presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2055]: Masha proposed using "single"+"single", but it's still not easy to implement, I have found another way, will post in our group chat later

Comment (xiaoxmeng) [reply presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2055]: Maybe put the solution in a github issue to ease review? Thanks!

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h [hunk 10]
```diff
@@ -187,6 +187,14 @@ class VeloxQueryPlanConverterBase {
   velox::VectorPtr evaluateConstantExpression(
       const velox::core::TypedExprPtr& expression);
 
+  void fillInAggregations(
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h:190]:
```
put input args first?
```

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 11]
```diff
@@ -1550,6 +1550,34 @@ velox::VectorPtr VeloxQueryPlanConverterBase::evaluateConstantExpression(
...
+  for (const auto& entry : outputVariables) {
+    aggregateNames.emplace_back(entry.name);
+    core::AggregationNode::Aggregate aggregate;
+    aggregate.call = std::dynamic_pointer_cast<const core::CallTypedExpr>(
+        exprConverter_.toVeloxExpr(aggregationMap.at(entry).call));
+    if (aggregationMap.at(entry).mask != nullptr) {
+      aggregate.mask =
+          exprConverter_.toVeloxExpr(aggregationMap.at(entry).mask);
+    }
+    if (aggregationMap.at(entry).orderBy) {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:1569]:
```
if (aggregationMap.at(entry).orderBy != nullptr) {
```

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 12]
```diff
@@ -2005,19 +2018,57 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
-       node->fragmentVariable,
-       node->tableCommitContextVariable});
+  std::vector<protocol::VariableReferenceExpression> variables = {
+      node->rowCountVariable,
+      node->fragmentVariable,
+      node->tableCommitContextVariable};
+
+  std::shared_ptr<core::AggregationNode> aggregationNode;
+  if (node->statisticsAggregation != nullptr) {
+    auto outputVariables = node->statisticsAggregation->outputVariables;
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2028]:
```
s/fillInAggregations/aggregationOutputVariables/
```

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 13]
```diff
@@ -2005,19 +2018,57 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
+        variables.end(), groupingVariables.begin(), groupingVariables.end());
+    for (auto const& variable : outputVariables) {
+      variables.push_back(variable);
+    }
+
+    std::vector<std::string> aggregateNames;
+    std::vector<core::AggregationNode::Aggregate> aggregates;
+    aggregateNames.reserve(aggregationMap.size());
+    aggregates.reserve(aggregationMap.size());
+    fillInAggregations(
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2048]: nit: we can handle the reserve inside fillInAggregations

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 14]
```diff
@@ -1630,30 +1658,15 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
-      for (const auto& orderBy : entry.second.orderBy->orderBy) {
-        aggregate.sortingKeys.emplace_back(
-            exprConverter_.toVeloxExpr(orderBy.variable));
-        aggregate.sortingOrders.emplace_back(
-            toVeloxSortOrder(orderBy.sortOrder));
-      }
-    }
-
-    aggregates.emplace_back(aggregate);
+  std::vector<protocol::VariableReferenceExpression> outputVariables;
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:1663]: Thanks for the refactoring!

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 15]
```diff
@@ -2005,19 +2018,57 @@ VeloxQueryPlanConverterBase::toVeloxQueryPlan(
...
+  auto outputType = toRowType(variables);
   return std::make_shared<core::TableWriteNode>(
       node->id,
       toRowType(node->columns),
       node->columnNames,
       insertTableHandle,
       outputType,
       getCommitStrategy(),
-      nullptr,
+      aggregationNode,
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2071]: std::move(aggregationNode)

**Review by xiaoxmeng — COMMENTED** [review]
@kewang1024 thanks for the update % minors.

**Review by xiaoxmeng — APPROVED** [review]
@kewang1024 LGTM. thanks for the update.

**Review by gggrace14 — APPROVED** [review]
Stamp

**Review by ajaygeorge — APPROVED** [review]

**Comment by kewang1024** [comment]
Followup e2e test: https://github.com/prestodb/presto/pull/21169

**Pull request merged**